Core/include/Acts/Vertexing/NumericalTrackLinearizer.ipp

```diff
@@ -0,0 +1,154 @@
+  pOptions.direction =
+      Direction::fromScalarZeroAsPositive(intersection.intersection.pathLength);
+
+  // Propagate to the PCA of linPointPos
+  auto result = m_cfg.propagator->propagate(params, *perigeeSurface, pOptions);
+
+  // Extracting the Perigee representation of the track wrt linPointPos
+  auto endParams = *result->endParameters;
+  BoundVector perigeeParams = endParams.parameters();
+
+  // Covariance and weight matrix at the PCA to "linPoint"
+  BoundSymMatrix parCovarianceAtPCA = endParams.covariance().value();
+  BoundSymMatrix weightAtPCA = parCovarianceAtPCA.inverse();
+
+  // Vector containing the track parameters at the PCA
+  // Note that we parametrize the track using the following parameters:
+  // (x, y, z, t, phi, theta, q/p),
+  // where
+  // -) (x, y, z, t) is the global 4D position of the PCA
+  // -) phi and theta are the global angles of the momentum at the PCA
+  // -) q/p is the charge divided by the total momentum at the PCA
+  const unsigned int nParams = 7;
+  Acts::ActsVector<nParams> paramVec;
+
+  // 4D PCA and the momentum of the track at the PCA
+  // These quantities will be used in the computation of the constant term in
+  // the Taylor expansion
+  Vector4 pca;
+  Vector3 momentumAtPCA;
+
+  // Fill "paramVec", "pca", and "momentumAtPCA"
+  {
+    Vector3 globalCoords = endParams.position(gctx);
+    ActsScalar globalTime = endParams.time();
+    ActsScalar phi = perigeeParams(BoundIndices::eBoundPhi);
+    ActsScalar theta = perigeeParams(BoundIndices::eBoundTheta);
+    ActsScalar qOvP = perigeeParams(BoundIndices::eBoundQOverP);
+
+    paramVec << globalCoords, globalTime, phi, theta, qOvP;
+    pca << globalCoords, globalTime;
+    momentumAtPCA << phi, theta, qOvP;
+  }
+
+  // Complete Jacobian (consists of positionJacobian and momentumJacobian)
+  ActsMatrix<eBoundSize, nParams> completeJacobian;
+  completeJacobian.setZero();
+
+  // Perigee parameters wrt linPoint after wiggling
+  BoundVector newPerigeeParams;
+
+  // Check if wiggled angle theta are within definition range [0, pi]
+  assert(paramVec(5) + m_cfg.delta < M_PI &&
+         "Wiggled theta outside range, choose a smaller wiggle (i.e., delta)!"
+         "You might need to decrease targetTolerance as well.");
+  // Wiggling each of the parameters at the PCA and computing the Perigee
+  // parametrization of the resulting new track. This allows us to approximate
+  // the numerical derivatives.
+  for (unsigned int i = 0; i < nParams; i++) {
+    // Wiggle
+    paramVec(i) += m_cfg.delta;
+
+    // Create curvilinear track object from our parameters. This is needed for
+    // the propagation. Note that we work without covariance since we don't need
+    // it to compute the derivative.
+    Vector3 wiggledDir =
+        makeDirectionUnitFromPhiTheta(paramVec(4), paramVec(5));
+    CurvilinearTrackParameters wiggledCurvilinearParams(
+        paramVec.head(4), wiggledDir, paramVec(6));
```

Comment:
might be better to use enums or even define these special track params in a more central way